File: sdks/python/kavachos/agents.py
"""Agent identity CRUD operations."""

from __future__ import annotations

from typing import List, Optional
from urllib.parse import quote

from kavachos._http import AsyncTransport, SyncTransport
from kavachos.types import (
    Agent,
    AgentFilters,
    AuthorizeRequest,
    AuthorizeResult,
    CreateAgentInput,
    Permission,
    UpdateAgentInput,
)


class AsyncAgentsResource:
    """Async agent identity management."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._http = transport

    async def create(self, input: CreateAgentInput) -> Agent:
        """Create a new agent identity.

        Args:
            input: Agent creation parameters.

        Returns:
            The newly created :class:`~kavachos.types.Agent`.
        """
        data = await self._http.request("POST", "/agents", json=input.to_dict())
        return Agent.from_dict(data)

    async def list(self, filters: Optional[AgentFilters] = None) -> List[Agent]:
        """List agent identities, optionally filtered.

        Args:
            filters: Optional query filters (user_id, status, type).

        Returns:
            List of :class:`~kavachos.types.Agent` objects.
        """
        params = filters.to_params() if filters else None
        data = await self._http.request("GET", "/agents", params=params)
        return [Agent.from_dict(a) for a in data]

    async def get(self, agent_id: str) -> Optional[Agent]:
        """Get an agent by ID. Returns ``None`` if not found.

        Args:
            agent_id: The agent's unique identifier.
        """
        from kavachos.errors import NotFoundError

        try:
            data = await self._http.request("GET", f"/agents/{quote(agent_id)}")
            return Agent.from_dict(data)
        except NotFoundError:
            return None

    async def update(self, agent_id: str, input: UpdateAgentInput) -> Agent:
        """Update an existing agent identity.

        Args:
            agent_id: The agent's unique identifier.
            input: Fields to update (all optional).

        Returns:
            The updated :class:`~kavachos.types.Agent`.
        """
        data = await self._http.request(
            "PATCH",
            f"/agents/{quote(agent_id)}",
            json=input.to_dict(),
        )
        return Agent.from_dict(data)

    async def revoke(self, agent_id: str) -> None:
        """Revoke (delete) an agent identity.

        Args:
            agent_id: The agent's unique identifier.
        """
        await self._http.request("DELETE", f"/agents/{quote(agent_id)}")

    async def rotate(self, agent_id: str) -> Agent:
        """Rotate an agent's bearer token.

        Issues a new cryptographic token and invalidates the previous one.

        Args:
            agent_id: The agent's unique identifier.

        Returns:
            The updated :class:`~kavachos.types.Agent` with a fresh token.
        """
        data = await self._http.request("POST", f"/agents/{quote(agent_id)}/rotate")
        return Agent.from_dict(data)

    async def authorize(self, agent_id: str, request: AuthorizeRequest) -> AuthorizeResult:
        """Check whether an agent is authorized to perform an action.

        Args:
            agent_id: The agent's unique identifier.
            request: The action and resource to check.

        Returns:
            :class:`~kavachos.types.AuthorizeResult` with ``allowed`` flag and audit ID.
        """
        data = await self._http.request(
            "POST",
            f"/agents/{quote(agent_id)}/authorize",
            json=request.to_dict(),
        )
        return AuthorizeResult.from_dict(data)

    async def get_effective_permissions(self, agent_id: str) -> List[Permission]:
        """Get the effective permissions for an agent, including delegated ones.

        Args:
            agent_id: The agent's unique identifier.

        Returns:
            Flattened list of :class:`~kavachos.types.Permission` objects.
        """
        data = await self._http.request(
            "GET",
            f"/delegations/{quote(agent_id)}/permissions",
        )
        return [Permission.from_dict(p) for p in data]


class SyncAgentsResource:
    """Sync agent identity management."""

    def __init__(self, transport: SyncTransport) -> None:
        self._http = transport

    def create(self, input: CreateAgentInput) -> Agent:
        """Create a new agent identity."""
        data = self._http.request("POST", "/agents", json=input.to_dict())
        return Agent.from_dict(data)

    def list(self, filters: Optional[AgentFilters] = None) -> List[Agent]:
        """List agent identities, optionally filtered."""
        params = filters.to_params() if filters else None
        data = self._http.request("GET", "/agents", params=params)
        return [Agent.from_dict(a) for a in data]

    def get(self, agent_id: str) -> Optional[Agent]:
        """Get an agent by ID. Returns ``None`` if not found."""
        from kavachos.errors import NotFoundError

        try:
            data = self._http.request("GET", f"/agents/{quote(agent_id)}")
            return Agent.from_dict(data)
        except NotFoundError:
            return None

    def update(self, agent_id: str, input: UpdateAgentInput) -> Agent:
        """Update an existing agent identity."""
        data = self._http.request(
            "PATCH",
            f"/agents/{quote(agent_id)}",
            json=input.to_dict(),
        )
        return Agent.from_dict(data)

    def revoke(self, agent_id: str) -> None:
        """Revoke (delete) an agent identity."""
        self._http.request("DELETE", f"/agents/{quote(agent_id)}")

    def rotate(self, agent_id: str) -> Agent:
        """Rotate an agent's bearer token."""
        data = self._http.request("POST", f"/agents/{quote(agent_id)}/rotate")
        return Agent.from_dict(data)

    def authorize(self, agent_id: str, request: AuthorizeRequest) -> AuthorizeResult:
        """Check whether an agent is authorized to perform an action."""
        data = self._http.request(
            "POST",
            f"/agents/{quote(agent_id)}/authorize",
            json=request.to_dict(),
        )
        return AuthorizeResult.from_dict(data)

    def get_effective_permissions(self, agent_id: str) -> List[Permission]:
        """Get the effective permissions for an agent, including delegated ones."""
        data = self._http.request(
            "GET",
            f"/delegations/{quote(agent_id)}/permissions",
        )
        return [Permission.from_dict(p) for p in data]

File: sdks/go/agents.go
package kavachos

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
)

// AgentsResource manages agent identity operations.
type AgentsResource struct {
	tp *transport
}

// Create creates a new agent identity.
func (r *AgentsResource) Create(ctx context.Context, input CreateAgentInput) (*Agent, error) {
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "POST", "/agents", input, nil, &raw); err != nil {
		return nil, err
	}
	var agent Agent
	if err := json.Unmarshal(unwrapAPIData(raw), &agent); err != nil {
		return nil, fmt.Errorf("kavachos: decode agent: %w", err)
	}
	return &agent, nil
}

// List returns agent identities. Pass nil filters to list all agents.
func (r *AgentsResource) List(ctx context.Context, filters *AgentFilters) ([]Agent, error) {
	params := agentFiltersToParams(filters)
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "GET", "/agents", nil, params, &raw); err != nil {
		return nil, err
	}
	data := unwrapAPIData(raw)
	var agents []Agent
	if err := json.Unmarshal(data, &agents); err != nil {
		return nil, fmt.Errorf("kavachos: decode agents list: %w", err)
	}
	return agents, nil
}

// Get retrieves an agent by ID. Returns nil, nil when not found.
func (r *AgentsResource) Get(ctx context.Context, agentID string) (*Agent, error) {
	var raw json.RawMessage
	err := r.tp.doJSON(ctx, "GET", "/agents/"+url.PathEscape(agentID), nil, nil, &raw)
	if err != nil {
		if IsNotFound(err) {
			return nil, nil
		}
		return nil, err
	}
	var agent Agent
	if err := json.Unmarshal(unwrapAPIData(raw), &agent); err != nil {
		return nil, fmt.Errorf("kavachos: decode agent: %w", err)
	}
	return &agent, nil
}

// Update modifies an existing agent. All fields in input are optional.
func (r *AgentsResource) Update(ctx context.Context, agentID string, input UpdateAgentInput) (*Agent, error) {
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "PATCH", "/agents/"+url.PathEscape(agentID), input, nil, &raw); err != nil {
		return nil, err
	}
	var agent Agent
	if err := json.Unmarshal(unwrapAPIData(raw), &agent); err != nil {
		return nil, fmt.Errorf("kavachos: decode agent: %w", err)
	}
	return &agent, nil
}

// Revoke deletes an agent identity.
func (r *AgentsResource) Revoke(ctx context.Context, agentID string) error {
	return r.tp.doJSON(ctx, "DELETE", "/agents/"+url.PathEscape(agentID), nil, nil, nil)
}

// Rotate issues a new bearer token for the agent and invalidates the previous one.
func (r *AgentsResource) Rotate(ctx context.Context, agentID string) (*Agent, error) {
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "POST", "/agents/"+url.PathEscape(agentID)+"/rotate", nil, nil, &raw); err != nil {
		return nil, err
	}
	var agent Agent
	if err := json.Unmarshal(unwrapAPIData(raw), &agent); err != nil {
		return nil, fmt.Errorf("kavachos: decode agent: %w", err)
	}
	return &agent, nil
}

// Authorize checks whether an agent is permitted to perform an action on a resource.
func (r *AgentsResource) Authorize(ctx context.Context, agentID string, req AuthorizeRequest) (*AuthorizeResult, error) {
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "POST", "/agents/"+url.PathEscape(agentID)+"/authorize", req, nil, &raw); err != nil {
		return nil, err
	}
	var result AuthorizeResult
	if err := json.Unmarshal(unwrapAPIData(raw), &result); err != nil {
		return nil, fmt.Errorf("kavachos: decode authorize result: %w", err)
	}
	return &result, nil
}

// GetEffectivePermissions returns the merged set of permissions for an agent,
// including any delegated permissions.
func (r *AgentsResource) GetEffectivePermissions(ctx context.Context, agentID string) ([]Permission, error) {
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "GET", "/delegations/"+url.PathEscape(agentID)+"/permissions", nil, nil, &raw); err != nil {
		return nil, err
	}
	var perms []Permission
	if err := json.Unmarshal(unwrapAPIData(raw), &perms); err != nil {
		return nil, fmt.Errorf("kavachos: decode permissions: %w", err)
	}
	return perms, nil
}

func agentFiltersToParams(f *AgentFilters) map[string]string {
	if f == nil {
		return nil
	}
	params := make(map[string]string)
	if f.UserID != nil {
		params["userId"] = *f.UserID
	}
	if f.Status != nil {
		params["status"] = string(*f.Status)
	}
	if f.Type != nil {
		params["type"] = string(*f.Type)
	}
	if len(params) == 0 {
		return nil
	}
	return params
}
